File: java/participant-config/Publisher.java
// java/participant-config -- publisher. Direct Java port of
// zig/participant-config/publisher.zig / c/participant-config/src/publisher.c
// / cpp/participant-config/src/publisher.cpp; see
// docs/design/participant-config-reference-app.md at the repo root for the
// full spec.
//
// Two mutually exclusive modes (see c/participant-config's publisher.c for
// the full description): programmatic (default) round-trips a
// Zzdds.zzdds.DomainParticipantConfig value through
// set_default_participant_config/get_default_participant_config before
// creating the participant via create_participant_ex; file mode
// (--config <path>) loads a zzdds.toml-style file instead. Both then run
// the same minimal reliable write loop (3 samples) as hello_world.
//
// Required stdout markers: "Create topic:", "Create writer for topic:",
// "on_reliable_reader_ready", "Publisher: wrote count=", "Publisher: done."
// Programmatic mode additionally prints "Config round-trip OK: ...". Any
// failure path prints a line starting "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;
import io.zzdds.ext.Zzdds;

import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

public class Publisher {
    static final int SAMPLE_COUNT = 3;
    static final int READER_READY_TIMEOUT_MS = 10000;
    static final int DRAIN_TIMEOUT_MS = 15000;
    static final int POLL_PERIOD_MS = 20;

    // Distinctive values for the programmatic round-trip check -- one plain
    // string field, one scalar field in an all-primitive nested struct, so a
    // mismatch in either is unambiguous about which kind of field broke.
    static final String CONFIG_PARTICIPANT_NAME = "participant-config-example";
    static final short CONFIG_FRAGMENT_SIZE = 9000;

    static class Options {
        int domainId = 0;
        String configPath = null;
    }

    static Options parseArgs(String[] args) {
        Options opts = new Options();
        for (int i = 0; i < args.length; i++) {
            if ((args[i].equals("-d") || args[i].equals("--domain")) && i + 1 < args.length) {
                opts.domainId = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--config") && i + 1 < args.length) {
                opts.configPath = args[++i];
            }
        }
        return opts;
    }

    public static void main(String[] args) throws Exception {
        Options opts = parseArgs(args);

        if (opts.configPath != null) {
            int cfgRc = io.zzdds.runtime.ZzddsRuntime.configureFromFile(opts.configPath);
            if (cfgRc != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: failed to load config file '" + opts.configPath + "' (rc=" + cfgRc + ")");
                System.exit(1);
            }
        }

        Object baseFactory = io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (baseFactory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp;
        if (opts.configPath == null) {
            // Programmatic mode: createFactory() always boxes into the base
            // DDS view -- narrow to zzdds's own extension view to reach
            // create_participant_ex/set_default_participant_config/
            // get_default_participant_config below.
            Zzdds.zzdds.DomainParticipantFactory factory =
                (Zzdds.zzdds.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.asZzddsFactory(baseFactory);
            if (factory == null) {
                System.err.println("FAIL: asZzddsFactory() failed");
                System.exit(1);
            }

            Zzdds.zzdds.DomainParticipantConfig cfg = new Zzdds.zzdds.DomainParticipantConfig();
            cfg.get_participant().set_name(CONFIG_PARTICIPANT_NAME);
            cfg.get_rtps().set_fragment_size(CONFIG_FRAGMENT_SIZE);

            if (factory.set_default_participant_config(cfg) != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: set_default_participant_config() failed");
                System.exit(1);
            }

            Zzdds.zzdds.DomainParticipantConfig readback = new Zzdds.zzdds.DomainParticipantConfig();
            if (factory.get_default_participant_config(readback) != Dcps.DDS.RETCODE_OK.value) {
                System.err.println("FAIL: get_default_participant_config() failed");
                System.exit(1);
            }

            if (!CONFIG_PARTICIPANT_NAME.equals(readback.get_participant().get_name())) {
                System.err.println("FAIL: participant.name round-trip mismatch: expected '" + CONFIG_PARTICIPANT_NAME
                        + "', got '" + readback.get_participant().get_name() + "'");
                System.exit(1);
            }
            if (readback.get_rtps().get_fragment_size() != CONFIG_FRAGMENT_SIZE) {
                System.err.println("FAIL: rtps.fragment_size round-trip mismatch: expected " + CONFIG_FRAGMENT_SIZE
                        + ", got " + readback.get_rtps().get_fragment_size());
                System.exit(1);
            }
            System.out.println("Config round-trip OK: participant.name='" + readback.get_participant().get_name()
                    + "' rtps.fragment_size=" + readback.get_rtps().get_fragment_size());

            dp = factory.create_participant_ex(opts.domainId, null, null, 0, cfg);
        } else {
            Dcps.DDS.DomainParticipantFactory factory = (Dcps.DDS.DomainParticipantFactory) baseFactory;
            dp = factory.create_participant(opts.domainId, null, null, 0);
        }
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + opts.domainId);
            System.exit(1);
        }

        if (ConfigPingTypeSupport.register(dp, "ConfigPing") != 0) {
            System.err.println("FAIL: register_type_support failed");
            System.exit(1);
        }

        Dcps.DDS.Topic topic = dp.create_topic("ConfigPing", "ConfigPing", null, null, 0);
        if (topic == null) {
            System.err.println("FAIL: create_topic() failed");
            System.exit(1);
        }
        System.out.println("Create topic: ConfigPing");

        Dcps.DDS.Publisher pub = dp.create_publisher(null, null, 0);
        if (pub == null) {
            System.err.println("FAIL: create_publisher() failed");
            System.exit(1);
        }

        Dcps.DDS.DataWriterQos dwQos = new Dcps.DDS.DataWriterQos();
        pub.get_default_datawriter_qos(dwQos);
        dwQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        dwQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        Dcps.DDS.DataWriter rawWriter = pub.create_datawriter(topic, dwQos, null, 0);
        if (rawWriter == null) {
            System.err.println("FAIL: create_datawriter() failed");
            System.exit(1);
        }
        System.out.println("Create writer for topic: ConfigPing");

        final AtomicBoolean readerReady = new AtomicBoolean(false);
        final AtomicBoolean everMatched = new AtomicBoolean(false);
        final AtomicInteger matchedCurrentCount = new AtomicInteger(0);

        Zzdds.zzdds.DataWriter zdWriter =
            (Zzdds.zzdds.DataWriter) io.zzdds.runtime.ZzddsRuntime.asZzddsDataWriter(rawWriter);
        if (zdWriter == null) {
            System.err.println("FAIL: asZzddsDataWriter() failed");
            System.exit(1);
        }

        Zzdds.zzdds.DataWriterListenerEx writerListener = new Zzdds.zzdds.DataWriterListenerEx() {
            public void on_offered_deadline_missed(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedDeadlineMissedStatus s) {}
            public void on_offered_incompatible_qos(Dcps.DDS.DataWriter w, Dcps.DDS.OfferedIncompatibleQosStatus s) {}
            public void on_liveliness_lost(Dcps.DDS.DataWriter w, Dcps.DDS.LivelinessLostStatus s) {}

            public void on_publication_matched(Dcps.DDS.DataWriter w, Dcps.DDS.PublicationMatchedStatus s) {
                matchedCurrentCount.set(s.get_current_count());
                if (s.get_current_count() > 0) everMatched.set(true);
                System.out.println("on_publication_matched() current_count=" + s.get_current_count());
            }

            public void on_reliable_reader_ready(int readerHandle, boolean isReady) {
                if (isReady) readerReady.set(true);
                System.out.println("on_reliable_reader_ready() is_ready=" + isReady);
            }
        };
        if (zdWriter.set_listener_ex(writerListener, Dcps.DDS.PUBLICATION_MATCHED_STATUS.value) != 0) {
            System.err.println("FAIL: set_listener_ex failed");
            System.exit(1);
        }

        ConfigPingDataWriter writer = new ConfigPingDataWriter(rawWriter);

        long deadline = System.currentTimeMillis() + READER_READY_TIMEOUT_MS;
        while (!readerReady.get()) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: no reliable reader became ready within " + (READER_READY_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        for (int i = 0; i < SAMPLE_COUNT; i++) {
            Config_ping.ConfigPing sample = new Config_ping.ConfigPing();
            sample.set_count(i);

            if (writer.write(sample, 0L) != 0) {
                System.err.println("FAIL: write() failed at count=" + i);
                System.exit(1);
            }
            System.out.println("Publisher: wrote count=" + i);
        }

        deadline = System.currentTimeMillis() + DRAIN_TIMEOUT_MS;
        while (!(everMatched.get() && matchedCurrentCount.get() == 0)) {
            if (System.currentTimeMillis() > deadline) {
                System.err.println("FAIL: subscriber did not disconnect within " + (DRAIN_TIMEOUT_MS / 1000) + "s");
                System.exit(1);
            }
            Thread.sleep(POLL_PERIOD_MS);
        }

        System.out.println("Publisher: done.");
    }
}

File: cpp/participant-config/src/publisher.cpp
/*
 * cpp/participant-config -- publisher. Direct C++ port of
 * zig/participant-config/publisher.zig / c/participant-config/src/publisher.c;
 * see docs/design/participant-config-reference-app.md at the repo root for
 * the full spec.
 *
 * Two mutually exclusive modes (see c/participant-config's publisher.c for
 * the full description): programmatic (default) round-trips a
 * zzdds::DomainParticipantConfig value through
 * set_default_participant_config/get_default_participant_config before
 * creating the participant via create_participant_ex; file mode
 * (--config <path>) loads a zzdds.toml-style file instead. Both then run
 * the same minimal reliable write loop (3 samples) as hello_world.
 *
 * Required stdout markers: "Create topic:", "Create writer for topic:",
 * "on_reliable_reader_ready", "Publisher: wrote count=", "Publisher: done."
 * Programmatic mode additionally prints "Config round-trip OK: ...". Any
 * failure path prints a line starting "FAIL:" and exits nonzero.
 */
#include "config_ping.hpp"
#include "zzdds_cpp.hpp"
#include "dcps_impl.hpp"

#include <atomic>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <memory>
#include <unistd.h>

namespace {

constexpr int SAMPLE_COUNT = 3;
constexpr int READER_READY_TIMEOUT_MS = 10000;
constexpr int DRAIN_TIMEOUT_MS = 15000;
constexpr int POLL_PERIOD_MS = 20;

// Distinctive values for the programmatic round-trip check -- one plain
// string field, one scalar field in an all-primitive nested struct, so a
// mismatch in either is unambiguous about which kind of field broke.
const char *CONFIG_PARTICIPANT_NAME = "participant-config-example";
constexpr uint16_t CONFIG_FRAGMENT_SIZE = 9000;

struct PubState {
    std::atomic<bool> reader_ready{false};
    std::atomic<bool> ever_matched{false};
    std::atomic<int> matched_current_count{0};
};

class PubListener : public ::zzdds::DataWriterListenerExBase {
public:
    explicit PubListener(PubState *state) : state_(state) {}

    void on_publication_matched(std::shared_ptr<::DDS::DataWriter> /*writer*/,
                                 ::DDS::PublicationMatchedStatus status) override {
        state_->matched_current_count.store(status.current_count);
        if (status.current_count > 0) state_->ever_matched.store(true);
        std::printf("on_publication_matched() current_count=%d\n", status.current_count);
    }

    void on_reliable_reader_ready(::DDS::InstanceHandle_t /*reader_handle*/, bool is_ready) override {
        if (is_ready) state_->reader_ready.store(true);
        std::printf("on_reliable_reader_ready() is_ready=%s\n", is_ready ? "true" : "false");
    }

private:
    PubState *state_;
};

struct Options {
    uint32_t domain_id = 0;
    const char *config_path = nullptr;
};

Options parse_args(int argc, char **argv) {
    Options opts;
    for (int i = 1; i < argc; i++) {
        if ((std::strcmp(argv[i], "-d") == 0 || std::strcmp(argv[i], "--domain") == 0) && i + 1 < argc) {
            opts.domain_id = static_cast<uint32_t>(std::strtoul(argv[++i], nullptr, 10));
        } else if (std::strcmp(argv[i], "--config") == 0 && i + 1 < argc) {
            opts.config_path = argv[++i];
        }
    }
    return opts;
}

} // namespace

int main(int argc, char **argv) {
    Options opts = parse_args(argc, argv);

    if (opts.config_path) {
        auto cfg_rc = zzdds::process_configure_from_file(opts.config_path, nullptr);
        if (cfg_rc != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: failed to load config file '%s' (rc=%d)\n", opts.config_path, static_cast<int>(cfg_rc));
            return 1;
        }
    }

    auto factory = zzdds::create_factory();
    if (!factory) {
        std::fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }

    std::shared_ptr<::DDS::DomainParticipant> dp;
    if (!opts.config_path) {
        // Programmatic mode.
        auto cfg = ::zzdds::DomainParticipantConfig::default_value();
        cfg.participant.name = CONFIG_PARTICIPANT_NAME;
        cfg.rtps.fragment_size = CONFIG_FRAGMENT_SIZE;

        if (factory->set_default_participant_config(cfg) != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: set_default_participant_config() failed\n");
            return 1;
        }

        auto readback = ::zzdds::DomainParticipantConfig::default_value();
        if (factory->get_default_participant_config(readback) != ::DDS::RETCODE_OK) {
            std::fprintf(stderr, "FAIL: get_default_participant_config() failed\n");
            return 1;
        }

        if (readback.participant.name != CONFIG_PARTICIPANT_NAME) {
            std::fprintf(stderr, "FAIL: participant.name round-trip mismatch: expected '%s', got '%s'\n",
                         CONFIG_PARTICIPANT_NAME, readback.participant.name.c_str());
            return 1;
        }
        if (readback.rtps.fragment_size != CONFIG_FRAGMENT_SIZE) {
            std::fprintf(stderr, "FAIL: rtps.fragment_size round-trip mismatch: expected %u, got %u\n",
                         static_cast<unsigned>(CONFIG_FRAGMENT_SIZE), static_cast<unsigned>(readback.rtps.fragment_size));
            return 1;
        }
        std::printf("Config round-trip OK: participant.name='%s' rtps.fragment_size=%u\n",
                    readback.participant.name.c_str(), static_cast<unsigned>(readback.rtps.fragment_size));

        dp = factory->create_participant_ex(opts.domain_id, ::DDS::DomainParticipantQos::default_value(), nullptr, 0, cfg);
    } else {
        dp = factory->create_participant(opts.domain_id, ::DDS::DomainParticipantQos::default_value(), nullptr, 0);
    }
    if (!dp) {
        std::fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", opts.domain_id);
        return 1;
    }
    auto dp_handle = dp->native_handle();

    if (ConfigPingTypeSupport::register_type(dp_handle) != 0) {
        std::fprintf(stderr, "FAIL: register_type failed\n");
        return 1;
    }

    auto topic = dp->create_topic("ConfigPing", "ConfigPing", ::DDS::TopicQos::default_value(), nullptr, 0);
    if (!topic) {
        std::fprintf(stderr, "FAIL: create_topic() failed\n");
        return 1;
    }
    std::printf("Create topic: ConfigPing\n");

    auto pub = dp->create_publisher(::DDS::PublisherQos::default_value(), nullptr, 0);
    if (!pub) {
        std::fprintf(stderr, "FAIL: create_publisher() failed\n");
        return 1;
    }

    auto dw_qos = ::DDS::DataWriterQos::default_value();
    dw_qos.reliability.kind = ::DDS::ReliabilityQosPolicyKind::RELIABLE_RELIABILITY_QOS;
    dw_qos.history.kind = ::DDS::HistoryQosPolicyKind::KEEP_ALL_HISTORY_QOS;

    auto dw = pub->create_datawriter(topic, dw_qos, nullptr, 0);
    if (!dw) {
        std::fprintf(stderr, "FAIL: create_datawriter() failed\n");
        return 1;
    }
    std::printf("Create writer for topic: ConfigPing\n");

    PubState state;
    auto listener = std::make_shared<PubListener>(&state);
    auto zdw = std::static_pointer_cast<::zzdds::DataWriterImpl>(dw);
    if (zdw->set_listener_ex(listener, DDS_PUBLICATION_MATCHED_STATUS) != ::DDS::RETCODE_OK) {
        std::fprintf(stderr, "FAIL: set_listener_ex failed\n");
        return 1;
    }

    auto dw_handle = dw->native_handle();
    ConfigPingDataWriter writer(dw_handle);

    for (int waited_ms = 0; !state.reader_ready.load(); waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= READER_READY_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: no reliable reader became ready within %ds\n", READER_READY_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    for (int i = 0; i < SAMPLE_COUNT; i++) {
        ::ConfigPing sample;
        sample.count = i;

        if (writer.write(sample) != 0) {
            std::fprintf(stderr, "FAIL: write() failed at count=%d\n", i);
            return 1;
        }
        std::printf("Publisher: wrote count=%d\n", i);
    }

    for (int waited_ms = 0;
         !(state.ever_matched.load() && state.matched_current_count.load() == 0);
         waited_ms += POLL_PERIOD_MS) {
        if (waited_ms >= DRAIN_TIMEOUT_MS) {
            std::fprintf(stderr, "FAIL: subscriber did not disconnect within %ds\n", DRAIN_TIMEOUT_MS / 1000);
            return 1;
        }
        usleep(POLL_PERIOD_MS * 1000);
    }

    std::printf("Publisher: done.\n");
    return 0;
}
